cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -438,7 +446,20 @@ pub(crate) fn put_input_in_rse_imm12<C: LowerCtx<I = Inst>>(
 ) -> ResultRSEImm12 {
     if let Some(imm_value) = input_to_const(ctx, input) {
         if let Some(i) = Imm12::maybe_from_u64(imm_value) {
-            return ResultRSEImm12::Imm12(i);
+            let out_ty_bits = ty_bits(ctx.output_ty(input.insn, 0));
+            let ty_signed_max = if out_ty_bits <= 1 {
+                out_ty_bits as i64
+            } else {
+                (1i64 << (out_ty_bits - 1)) - 1
+            };
+            let is_negative = i.bits > (ty_signed_max as u16);
```

Comment:
Can we just do this as: `let is_negative = i.bits & (1 << (out_ty_bits - 1)) != 0` (i.e., check the MSB / sign bit)?

Reply:
`iconst.i8 0xff00` should result in the non-negative value 0. There is no guarantee that the immediate is either zero or sign extended. Only the n least significant bits should be considered for an n bit big type.

Reply:
For clarity, @bjorn3, are you referring to the original or the suggested change? I think that checking the MSB only (`1 << 7` in the `i8` case) will correctly indicate that `iconst.i8 0xff00` is non-negative when masked to the proper width; whereas the `bits > ty_signed_max` logic will indicate it is negative.

Reply:
With your version it is ok, though I think you misplaced a paren. (`- 1))` instead kf `) -1)`)

Reply:
> Can we just do this as: let is_negative = i.bits & (1 << (out_ty_bits - 1)) != 0 (i.e., check the MSB / sign bit)?

Had to change it slightly because it was underflowing in `out_ty_bits`.

> `iconst.i8 0xff00` should result in the non-negative value 0

This one I feel like we should forbid? It feels like malformed input.

Reply:
>> iconst.i8 0xff00 should result in the non-negative value 0
>
> This one I feel like we should forbid? It feels like malformed input.

Hmm, this is actually interesting: in the absence of signedness (until use), there are two possible ranges (-128..127 or 0..255 in the `i8` case). If we allow the union of these, we have a 1.5*2^N-sized range, which is a little weird, but makes some sense. And, while it shouldn't be authoritative for our case, I was curious and just checked what `wat2wasm` does: it accepts exactly that range in the analogous case (I tested `i32.const` with -2^31..2^32-1), but roundtrips the values as signed unconditionally.

So, this feels like a reasonable behavior to me: we should validate in the CLIF validator that const values are in either signed or unsigned range for the type.

This can be a followup fix, though; no need to solve it here.

> With your version it is ok, though I think you misplaced a paren. (- 1)) instead of ) -1))

It's actually inside the paren correctly, I think: we want the next-lower power of two (bit 7 for an i8), not an all-ones mask. I think @afonso360's latest commit got this right fwiw.

Reply:
> This can be a followup fix, though; no need to solve it here.

Let me create an issue for this

Reply:
Thanks!